resharper/src/resharper-unity/CSharp/Daemon/Stages/Analysis/PreferGenericMethodOverloadAnalyzer.cs

```diff
@@ -1,202 +1,76 @@
+            // Don't add the quick fix in the case of the following, because we can't fix it cleanly:
+            // GetComponent(
+            // #if DEBUG
+            //    "MyDebugComponent"
+            // #else
+            //    "MyReleaseComponent"
+            // #endif
+            // )
+            if (expression.ContainsPreprocessorDirectives())
                 return;
-            }
-            
-            if (types.Length == 0)
-            {
-                consumer.AddHighlighting(new UnknownTypeInStringLiteralWarning(argument));
-                return;
-            }
-            
-            // Notify if ambiguous type, because it is unexpected behaviour. Unity does not specify which type will be chosen
-            // if two component with same name will be attached to object
-            if (typesWithRightInheritance.Length >= 2)
-            {
-                consumer.AddHighlighting(new AmbiguousTypeInStringLiteralWarning(argument));
-            }
-        }
 
-        private static string GetMethodName(IReference reference)
-        {
-            var info = reference.Resolve();
-            
-            if (info.ResolveErrorType == ResolveErrorType.OK)
-            {
-                return info.DeclaredElement?.ShortName;
-            }
-            return null;
-        }
-        
-        private static IClrTypeName GetContainingClrTypeName([NotNull] IReference reference)
-        {
-            var info = reference.Resolve();
-            
-            if (info.ResolveErrorType == ResolveErrorType.OK)
-            {
-                var method = info.DeclaredElement as IMethod;
-                return method?.GetContainingType()?.GetClrName();
-            }
-            return null;
-        }
-
-        [NotNull]
-        private ImportTypeResolver TypeResolverFactory([NotNull] ITreeNode context)
-        {
-            var symbolCache = context.GetPsiServices().Symbols;
-            var symbolScope = symbolCache.GetSymbolScope(context.GetPsiModule(), withReferences: true, true);
-            
-            return typeName => symbolScope.GetElementsByShortName(typeName);
-        }
+            var references = literalExpressionArgument.GetReferences<UnityObjectTypeOrNamespaceReference>();
 
-        private ITypeElement[] ResolveStringLiteral(string literal, ITreeNode context)
-        {
-            var factory = CSharpElementFactory.GetInstance(context);
-            
-            var typeNode = factory.CreateTypeUsageNode(literal, context);
-            var typeReference = ((IReferenceName)typeNode.FirstChild)?.Reference;
-            if (typeReference == null)
+            // Don't add anything unless ALL references resolve properly
+            foreach (var reference in references)
             {
-                return Array.Empty<ITypeElement>();
+                var resolveInfo = reference.Resolve();
+                if (!resolveInfo.Info.ResolveErrorType.IsAcceptable)
```

Comment:
In this case, we will never show quick fix with possibility to choose specified class in case with ambiguous classes
````
namespace A 
{
    class Test  : MonoBehaviour
    {
    }
}

namespace B
{
   class Test : MonoBehaviour 
   {
   }
}

class Example : MonoBehaviour 
{
    void Start() 
    {
        var test = GetComponent("Test");
    }
}

```

Reply:
The reference resolve will show a warning, with a quick fix to choose which one. Selecting that will change it to a fully qualified string, and then we get the quick fix to rewrite as generic method. I tried it first with the quick fix on multiple candidates, but the interaction is weird - the warning for the string literal overwrites the ambiguous match warning, which I think is more important.